### PR TITLE
Add per-event timezone support to The Events Calendar

### PR DESCRIPTION
**Pull request by barryhughes**
For context [please see #37395](https://central.tri.be/issues/37395). Some general notes:

* Timezone settings (in *Events > Settings > General)* don't become available until all events have been updated with timezone data 
* In the absence of any other data (ie, pre-update), the existing `_EventStartDate` is taken to be in the "local event timezone" 
* If not specifically set, the local event timezone is taken to be the same as the WP timezone

### CODE REVIEW

File: src/Tribe/Admin/Timezone_Settings.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
+<?php
+class Tribe__Events__Admin__Timezone_Settings {
+
+
+	public function __construct() {
```
Comment (borkweb) [inline src/Tribe/Admin/Timezone_Settings.php]: Docblock comments would be good to have on all of these methods.

File: src/Tribe/Admin/Timezone_Settings.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+	function ajax_updater() {
+		if ( ! isset( $_POST['check'] ) || ! wp_verify_nonce( $_POST['check'], 'timezone-settings' ) ) {
+			return;
+		}
+
+		$updater = new Tribe__Events__Admin__Timezone_Updater;
+		$updater->init_update();
+
+		exit( json_encode( array(
```
Comment (borkweb) [inline src/Tribe/Admin/Timezone_Settings.php]: Perhaps use [`wp_send_json`](https://github.com/WordPress/WordPress/blob/master/wp-includes/functions.php#L2801) instead?

File: src/Tribe/Admin/Timezone_Settings.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+			'html'     => $updater->notice_inner(),
+			'continue' => $updater->update_needed()
```
Comment (borkweb) [inline src/Tribe/Admin/Timezone_Settings.php]: Add a trailing comma

File: src/Tribe/Admin/Timezone_Settings.php (hunk 2 above)
Comment (barryhughes) [inline src/Tribe/Admin/Timezone_Settings.php]: :+1: ... have to admit I was until now completely ignorant of that function.

File: src/Tribe/Admin/Timezone_Updater.php [hunk 4]
```diff
@@ -0,0 +1,186 @@
+<?php
+/**
+ * Tool for adding timezone data to events.
+ *
+ * The application for this is in transitioning any event data created in 3.11.x or
+ * earlier that hasn't since been updated, so that it becomes "timezone ready".
+ */
+class Tribe__Events__Admin__Timezone_Updater {
```
Comment (borkweb) [inline src/Tribe/Admin/Timezone_Updater.php]: I'm a fan of this class. Nice work here!

File: src/Tribe/Admin/Timezone_Updater.php [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	protected $initial_count = 0;
+
+	public function init_update() {
```
Comment (borkweb) [inline src/Tribe/Admin/Timezone_Updater.php]: Moar docblock comments :)

File: src/Tribe/Settings.php [hunk 6]
```diff
@@ -258,7 +258,7 @@ public function generatePage() {
 				wp_nonce_field( 'saving', 'tribe-save-settings' );
 				echo '<div class="clear"></div>';
 				echo '<input type="hidden" name="current-settings-tab" id="current-settings-tab" value="' . esc_attr( $this->currentTab ) . '" />';
-				echo '<input id="tribeSaveSettings" class="button-primary" type="submit" name="tribeSaveSettings" value="' . esc_attr__( 'Save Changes', 'tribe-events-calendar' ) . '" />';
+				echo '<input id="tribeSaveSettings" class="button-primary" type="submit" name="tribeSaveSettings" value="' . esc_attr__( ' Save Changes', 'tribe-events-calendar' ) . '" />';
```
Comment (borkweb) [inline src/Tribe/Settings.php]: We probably don't want that space there.

File: src/Tribe/Timezones.php [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+			$timezone_string = self::get_event_timezone_string( $event_id );
+			$date = tribe_get_start_date( $event_id, true, Tribe__Events__Date_Utils::DBDATETIMEFORMAT );
+			$abbr = self::abbr( $date, $timezone_string );
+		}
+
+		return empty( $abbr )
+			? $timezone_string
+			: $abbr;
+	}
+	/**
```
Comment (borkweb) [inline src/Tribe/Timezones.php]: Maybe add a linebreak between that function and the docblock

File: src/admin-views/tribe-options-timezones.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
+<?php
+defined( 'ABSPATH' ) or exit();
+
+$enable_button_label  = __( 'Enable timezone support', 'tribe-events-calendar' );
+$enable_button_url    = esc_url( wp_nonce_url( add_query_arg( 'timezone-update', '1', Tribe__Events__Settings::instance()->get_url() ), 'timezone-settings' ) );
+$enable_button_text   = __( 'Update Timezone Data', 'tribe-events-calendar' );
+$enable_button_help   = sprintf( __( 'Click this button to update your database and take advantage of additional timezone capabilities. Please <a href="%s" target="_blank">configure WordPress</a> to use the correct timezone before clicking this button!', 'tribe-events-calendar' ),
+	get_admin_url( null, 'options-general.php' )
```
Comment (borkweb) [inline src/admin-views/tribe-options-timezones.php]: `esc_url()` wrap the `get_admin_url()` call

File: src/functions/template-tags/date.php [hunk 9]
```diff
@@ -141,46 +133,31 @@ function tribe_get_start_date( $event = null, $displayTime = true, $dateFormat =
...
-			}
-			$date = strtotime( $event->EventEndDate );
-		} else {
-			return;
+		if ( tribe_event_is_all_day( $event ) ) {
+			$displayTime = false;
 		}
 
-		return tribe_event_format_date( $date, $displayTime, $dateFormat );
+		/* @todo revise for all-day-event past eod cutoff handling */
```
Comment (borkweb) [inline src/functions/template-tags/date.php]: Was this `@todo` addressed?

Comment (barryhughes) [inline src/functions/template-tags/date.php]: Yes - I'll remove it and add a specific testing point to the QA instructions.

File: src/Tribe/Timezones.php [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+		self::display_timezones();
+	}
+
+	/**
+	 * Takes care of appending timezone information to the display of
+	 * event date/times.
+	 */
+	protected static function display_timezones() {
+		if ( tribe_get_option( 'tribe_events_timezones_show_zone' ) ) {
+			add_filter( 'tribe_events_event_schedule_details', array( __CLASS__, 'append_timezone' ), 10, 2 );
```
Comment (borkweb) [inline src/Tribe/Timezones.php]: Appending the timezone using this filter causes the timezone to be inserted after the `$before`/`$after` values passed in to the `tribe_events_event_schedule_details()` function:

This (using one of my events on my local instance):

```
tribe_events_event_schedule_details( $event, '<h3>', '</h3>' );
```

Would result in:

```
<h3>
  <span class="date-start dtstart">
    August 6@3:00 pm
    <span class="value-title" title="2015-08-06PDT03:00"></span>
  </span>
  - 
  <span class="end-time dtend">
    8:00 pm
    <span class="value-title" title="2015-08-06PDT08:00"></span>
  </span>
</h3>
<span class="timezone"> PDT </span>
```

Which results in some funkiness with the appearance:

screen shot 2015-08-05 at 4 41 51 pm

Perhaps we need to add a new filter [around `$schedule`](https://github.com/moderntribe/the-events-calendar/blob/release/120/src/functions/template-tags/general.php#L1116) and use that filter instead?

Comment (zbtirrell) [inline src/Tribe/Timezones.php]: Any chance you want to knock this out while fiddling with that appearance? https://central.tri.be/issues/37270

Comment (borkweb) [inline src/Tribe/Timezones.php]: Is that the issue that you were thinking it was, @zbtirrell? It doesn't look related.

Comment (barryhughes) [inline src/Tribe/Timezones.php]: I was going to update the `tribe_events_event_schedule_details` hook so callbacks could access the before and after properties type-of-deal, rather than any CSS work.

Comment (zbtirrell) [inline src/Tribe/Timezones.php]: on a quick glance, I thought that was the countdown widget.  It's not... ignore me.

**Comment by borkweb** [comment]
:+1:
